_posts/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors.md

```diff
@@ -0,0 +1,155 @@
+
+### Accuracy
+
+In terms of recall, you can expect around 0.97 recall compared to exact search. OpenSearch uses the HNSW algorithm for approximate nearest neighbor searches, but accuracy depends on your dataset. Some models produce binary vectors with high accuracy: for example, [Cohere Embed v3](https://cohere.com/blog/int8-binary-embeddings) reports a 94.7% match in search quality compared to FP32 embeddings. Thus, when using a model producing quality binary embeddings, binary vectors can reach nearly the same accuracy as FP32 vectors.
+
+## Challenges with binary vectors: When they fall short
+
+When your model only produces FP32 vectors but you’d like to use binary vectors in OpenSearch, the process can get complicated. The following example demonstrates binary vector search in OpenSearch using FP32 vectors, along with the challenges involved.
+
+For this example, we used the [Cohere Simple dataset](https://huggingface.co/datasets/nreimers/wikipedia-22-12-large/tree/main) from Hugging Face. Since the data was in FP32, we converted it to binary by setting zero and negative values to `0`, and positive values to `1`. The following image illustrates the conversion process.
+
+<img src="/assets/media/blog-images/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors/pic4.png" alt="binary-quantization" class="img-centered"/>
+
+In terms of recall, binary vectors achieved a score of 0.73196.  To reach a recall of 0.93865, we needed 3x oversampling. Additionally, rescoring requires storing the original vector format, which adds disk usage unless the original vector is stored outside OpenSearch. The following image illustrates oversampling with rescoring. Note that recall may vary for different datasets.
+
+<img src="/assets/media/blog-images/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors/pic5.png" alt="oversampling-re-scoring" class="img-centered"/>
+
+By quantizing, oversampling, and rescoring, binary vectors can achieve similar recall to FP32 while using significantly less memory. However, managing these steps outside of OpenSearch can be cumbersome. [Disk-based vector search](https://opensearch.org/docs/latest/search-plugins/knn/disk-based-vector-search/) simplifies the process by automatically handling all required steps using advanced quantization techniques. Give it a try and see the difference!
+
+## Conclusion
+
+Binary vectors offer an efficient alternative to FP32 vectors, reducing memory and storage usage by over 90% while maintaining strong performance on smaller hardware. This efficiency makes binary vectors ideal for large-scale vector search applications like recommendation systems or search engines, where speed and resource savings are critical. If you’re handling massive datasets, binary vectors provide a practical way to scale search capabilities without increasing costs—--try them on your data to experience the difference.
+
```

Comment:
The text I deleted on line 147 was essentially a repeat of the last sentence on line 143.